README.md
# mlr3learners.coxboost

<!-- badges: start -->
[![tic](https://github.com/mlr3learners/mlr3learners.coxboost/workflows/tic/badge.svg?branch=master)](https://github.com/mlr3learners/mlr3learners.coxboost/actions)
[![R CMD Check via {tic}](https://github.com/mlr3learners/mlr3learners.coxboost/workflows/R%20CMD%20Check%20via%20{tic}/badge.svg?branch=master)](https://github.com/mlr3learners/mlr3learners.coxboost/actions)
![mlr3learners.drat](https://github.com/mlr3learners/mlr3learners.coxboost/workflows/mlr3learners.drat/badge.svg?branch=master)
[![Parameter Check](https://github.com/mlr3learners/mlr3learners.coxboost/workflows/Parameter%20Check/badge.svg?branch=master)](https://github.com/mlr3learners/mlr3learners.coxboost/actions)
[![codecov](https://codecov.io/gh/mlr3learners/mlr3learners.coxboost/branch/master/graph/badge.svg)](https://codecov.io/gh/mlr3learners/mlr3learners.coxboost)
[![StackOverflow](https://img.shields.io/badge/stackoverflow-mlr3-orange.svg)](https://stackoverflow.com/questions/tagged/mlr3)

<!-- badges: end -->

Adds `coxboost, cvcoxboost` from the {CoxBoost} package to {mlr3}.

Install the latest release of the package via

```r
install.packages("mlr3learners.coxboost")
```

by following the instructions in the [mlr3learners.drat README](https://github.com/mlr3learners/mlr3learners.drat).

Alternatively, you can install the latest version of {mlr3learners.coxboost} from Github with:

```r
remotes::install_github("mlr3learners/mlr3learners.coxboost")
```


NEWS.md
# mlr3learners.coxboost 0.1.2

- distr6 patch

# mlr3learners.coxboost 0.1.1

- Removed remotes dependencies
- Minor internal bug fix


# mlr3learners.coxboost 0.1.0

- Initial release


tests/testthat.R
library(testthat)
library(mlr3learners.coxboost)

test_check("mlr3learners.coxboost")


tests/testthat/test_surv_coxboost.R
context("surv.coxboost")

test_that("autotest", {
  learner = LearnerSurvCoxboost$new()
  expect_learner(learner)
  result = run_autotest(learner,check_replicable = FALSE)
  expect_true(result, info = result$error)
})


tests/testthat/test_surv_cvcoxboost.R
context("surv.cvcoxboost")

test_that("autotest", {
  set.seed(1)
  learner = lrn("surv.cvcoxboost", maxstepno = 1, K = 2)
  expect_learner(learner)
  result = run_autotest(learner, N = 10,check_replicable = FALSE)
  expect_true(result, info = result$error)
})

test_that("optim", {
  set.seed(1)
  learner = lrn("surv.cvcoxboost",
    penalty = "optimCoxBoostPenalty",
    maxstepno = 1, minstepno = 0, iter.max = 1, K = 2)
  task = tgen("simsurv")$generate(10)
  expect_prediction_surv(suppressWarnings(learner$train(task)$predict(task)))
})


tic.R
if (!ci_has_env("PARAMTEST") && !ci_has_env("DRAT")) {
  do_package_checks()

  # add random delay (up to 20s) to avoid git racing conditions
  get_stage("deploy") %>%
    add_code_step(floor(runif(15, min = 0, max = 20)))
  do_drat("mlr3learners/mlr3learners.drat")

} else if (ci_has_env("PARAMTEST")) {
  # PARAMTEST
  get_stage("install") %>%
    add_step(step_install_deps())

  get_stage("script") %>%
    add_code_step(remotes::install_dev("mlr3")) %>%
    add_code_step(testthat::test_dir(system.file("paramtest",
      package = "mlr3learners.coxboost"),
    stop_on_failure = TRUE))
} else if (ci_has_env("DRAT")) {


  get_stage("deploy") %>%
    add_step(step_install_deps()) %>%
    # add random delay (up to 5 mins) to avoid git racing conditions
    add_code_step(floor(runif(15, min = 0, max = 300))) %>%
    # remove devel version indicator to enable deployment
    add_code_step(writeLines(
      gsub(".9000", "", readLines("DESCRIPTION")),
      "DESCRIPTION"))
  do_drat("mlr3learners/mlr3learners.drat")
}


R/zzz.R
#' @import data.table
#' @import paradox
#' @import mlr3misc
#' @import survival
#' @importFrom R6 R6Class
#' @importFrom mlr3 mlr_learners LearnerClassif LearnerRegr
#' @importFrom mlr3proba LearnerSurv
"_PACKAGE"

# nocov start
register_mlr3 = function(libname, pkgname) {
  # get mlr_learners dictionary from the mlr3 namespace
  x = utils::getFromNamespace("mlr_learners", ns = "mlr3")

  # add the learner to the dictionary
  x$add("surv.coxboost", LearnerSurvCoxboost)
  x$add("surv.cvcoxboost", LearnerSurvCVCoxboost)
}

.onLoad = function(libname, pkgname) { # nolint
  register_mlr3()
  setHook(packageEvent("mlr3", "onLoad"), function(...) register_mlr3(),
    action = "append")
}

.onUnload = function(libpath) { # nolint
  event = packageEvent("mlr3", "onLoad")
  hooks = getHook(event)
  pkgname = vapply(hooks, function(x) environment(x)$pkgname, NA_character_)
  setHook(event, hooks[pkgname != "mlr3learners.coxboost"],
    action = "replace")
}
# nocov end


R/LearnerSurvCoxboost.R
#' @title Survival Cox Model with Likelihood Based Boosting Learner
#'
#' @name mlr_learners_surv.coxboost
#'
#' @description
#' A [mlr3proba::LearnerSurv] implementing coxboost from package
#'   \CRANpkg{CoxBoost}.
#' Calls [CoxBoost::CoxBoost()].
#'
#' @details
#' Use [LearnerSurvCoxboost] and [LearnerSurvCVCoxboost] for Cox boosting without and with internal
#' cross-validation of boosting step number, respectively. Tuning using the internal optimizer in
#' [LearnerSurvCVCoxboost] may be more efficient when tuning `stepno` only. However, for tuning
#' multiple hyperparameters, \CRANpkg{mlr3tuning} and [LearnerSurvCoxboost] will likely give better
#' results.
#'
#' @templateVar id surv.coxboost
#' @template section_dictionary_learner
#'
#' @references
#' Binder, H., Allignol, A., Schumacher, M., and Beyersmann, J. (2009).
#' Boosting for high-dimensional time-to-event data with competing risks.
#' Bioinformatics, 25:890-896.
#'
#' @template seealso_learner
#' @template example
#' @export
# <Adapt the name to your learner. For regression learners inherit = LearnerRegr>
LearnerSurvCoxboost = R6Class("LearnerSurvCoxboost",
  inherit = LearnerSurv,

  public = list(
    #' @description
    #' Creates a new instance of this [R6][R6::R6Class] class.
    initialize = function() {
      ps = ParamSet$new(
        params = list(
          ParamUty$new(id = "unpen.index", tags = "train"),
          ParamLgl$new(id = "standardize", default = TRUE, tags = "train"),
          ParamInt$new(id = "stepno", default = 100, lower = 0, tags = "train"),
          ParamDbl$new(id = "penalty", tags = "train"),
          ParamFct$new(
            id = "criterion", default = "pscore",
            levels = c("pscore", "score", "hpscore", "hscore"), tags = "train"),
          ParamDbl$new(id = "stepsize.factor", default = 1, tags = "train"),
          ParamFct$new(
            id = "sf.scheme", default = "sigmoid", levels = c("sigmoid", "linear"),
            tags = "train"),
          ParamUty$new(id = "pendistmat", tags = "train"),
          ParamUty$new(id = "connected.index", tags = "train"),
          ParamLgl$new(id = "x.is.01", default = FALSE, tags = "train"),
          ParamLgl$new(id = "return.score", default = TRUE, tags = "train"),
          ParamLgl$new(id = "trace", default = FALSE, tags = "train"),
          ParamUty$new(id = "at.step", tags = "predict")
      ))

      super$initialize(
        # see the mlr3book for a description: https://mlr3book.mlr-org.com/extending-mlr3.html
        id = "surv.coxboost",
        packages = "CoxBoost",
        feature_types = c("integer", "numeric", "factor", "logical"),
        predict_types = c("distr", "crank", "lp"),
        param_set = ps,
        properties = "weights",
        # the help file name is the one used as @name in the roxygen2 block
        man = "mlr3learners.coxboost::mlr_learners_surv.coxboost"
      )
    }


    # <Add method for importance, if learner supports that>
    # <See mlr3learners.randomForest for an example>

    # <Add method for oob_error, if learner supports that.>

  ),

  private = list(
    .train = function(task) {
      pars = self$param_set$get_values(tags = "train")

      if ("weights" %in% task$properties) {
        pars$weights = as.numeric(task$weights$weight)
      }

      with_package("CoxBoost", {
        mlr3misc::invoke(
          CoxBoost::CoxBoost,
          time = task$truth()[, 1],
          status = task$truth()[, 2],
          x = model.matrix(
            ~.,
            as.data.frame(task$data(cols = task$feature_names)))[, -1, drop = FALSE],
          .args = pars
        )
      })
    },

    .predict = function(task) {

      lp = as.numeric(mlr3misc::invoke(predict,
        self$model,
        newdata = model.matrix(
          ~.,
          as.data.frame(task$data(cols = task$feature_names)))[, -1,
          drop = FALSE],
        .args = self$param_set$get_values(tags = "predict"),
        type = "lp"))

      cdf = mlr3misc::invoke(predict,
        self$model,
        newdata = model.matrix(
          ~.,
          as.data.frame(task$data(cols = task$feature_names)))[, -1,
          drop = FALSE],
        .args = self$param_set$get_values(tags = "predict"),
        type = "CIF",
        times = sort(unique(self$model$time)))

      # define WeightedDiscrete distr6 object from predicted survival function
      x = rep(list(list(x = sort(unique(self$model$time)), cdf = 0)), task$nrow)
      for (i in 1:task$nrow) {
        x[[i]]$cdf = cdf[i, ]
      }

      distr = distr6::VectorDistribution$new(
        distribution = "WeightedDiscrete", params = x,
        decorators = c("CoreStatistics", "ExoticStatistics"))

      mlr3proba::PredictionSurv$new(task = task, crank = lp, distr = distr, lp = lp)
    }
  )
)


R/LearnerSurvCVCoxboost.R
#' @title Survival Cox Model with Cross-Validation Likelihood Based Boosting Learner
#'
#' @name mlr_learners_surv.cvcoxboost
#'
#' @description
#' A [mlr3proba::LearnerSurv] implementing cvcoxboost from package
#'   \CRANpkg{CoxBoost}.
#' Calls [CoxBoost::cv.CoxBoost()].
#'
#' @details
#' Use [LearnerSurvCoxboost] and [LearnerSurvCVCoxboost] for Cox boosting without and with internal
#' cross-validation of boosting step number, respectively. Tuning using the internal optimizer in
#' [LearnerSurvCVCoxboost] may be more efficient when tuning `stepno` only. However, for tuning
#' multiple hyperparameters, \CRANpkg{mlr3tuning} and [LearnerSurvCoxboost] will likely give better
#' results.
#'
#' If `penalty == "optimCoxBoostPenalty"` then [CoxBoost::optimCoxBoostPenalty] is used to determine
#' the penalty value to be used in [CoxBoost::cv.CoxBoost].
#'
#' @templateVar id surv.cvcoxboost
#' @template section_dictionary_learner
#'
#' @references
#' Binder, H., Allignol, A., Schumacher, M., and Beyersmann, J. (2009).
#' Boosting for high-dimensional time-to-event data with competing risks.
#' Bioinformatics, 25:890-896.
#'
#' @template seealso_learner
#' @template example
#' @export
# <Adapt the name to your learner. For regression learners inherit = LearnerRegr>
LearnerSurvCVCoxboost = R6Class("LearnerSurvCVCoxboost",
  inherit = LearnerSurv,

  public = list(
    #' @description
    #' Creates a new instance of this [R6][R6::R6Class] class.
    initialize = function() {
      ps = ParamSet$new(
        params = list(
          ParamInt$new(
            id = "maxstepno", default = 100, lower = 0,
            tags = c("train", "cvpars")),
          ParamInt$new(id = "K", default = 10, lower = 2, tags = c("train", "cvpars")),
          ParamFct$new(
            id = "type", default = "verweij", levels = c("verweij", "naive"),
            tags = c("train", "cvpars")),
          ParamUty$new(id = "folds", default = NULL, tags = c("train", "cvpars")),
          ParamInt$new(
            id = "minstepno", default = 50, lower = 0,
            tags = c("train", "optimPenalty")),
          ParamDbl$new(id = "start.penalty", tags = c("train", "optimPenalty")),
          ParamInt$new(id = "iter.max", default = 10, lower = 1, tags = c("train", "optimPenalty")),
          ParamDbl$new(
            id = "upper.margin", default = 0.05, lower = 0, upper = 1,
            tags = c("train", "optimPenalty")),
          ParamUty$new(id = "unpen.index", tags = "train"),
          ParamLgl$new(id = "standardize", default = TRUE, tags = "train"),
          ParamDbl$new(
            id = "penalty", special_vals = list("optimCoxBoostPenalty"),
            tags = c("train", "optimPenalty")),
          ParamFct$new(
            id = "criterion", default = "pscore",
            levels = c("pscore", "score", "hpscore", "hscore"), tags = "train"),
          ParamDbl$new(id = "stepsize.factor", default = 1, tags = "train"),
          ParamFct$new(
            id = "sf.scheme", default = "sigmoid", levels = c("sigmoid", "linear"),
            tags = "train"),
          ParamUty$new(id = "pendistmat", tags = "train"),
          ParamUty$new(id = "connected.index", tags = "train"),
          ParamLgl$new(id = "x.is.01", default = FALSE, tags = "train"),
          ParamLgl$new(id = "return.score", default = TRUE, tags = "train"),
          ParamLgl$new(id = "trace", default = FALSE, tags = "train"),
          ParamUty$new(id = "at.step", tags = "predict")
        )
      )

      super$initialize(
        # see the mlr3book for a description: https://mlr3book.mlr-org.com/extending-mlr3.html
        id = "surv.cvcoxboost",
        packages = "CoxBoost",
        feature_types = c("integer", "numeric", "factor", "logical"),
        predict_types = c("distr", "crank", "lp"),
        param_set = ps,
        properties = "weights",
        # the help file name is the one used as @name in the roxygen2 block
        man = "mlr3learners.coxboost::mlr_learners_surv.cvcoxboost"
      )
    }
  ),

  private = list(
    .train = function(task) {

      opt_pars = self$param_set$get_values(tags = "optimPenalty")
      cv_pars = self$param_set$get_values(tags = "cvpars")
      cox_pars = setdiff(
        self$param_set$get_values(tags = "train"),
        c(opt_pars, cv_pars))

      if ("weights" %in% task$properties) {
        cox_pars$weights = as.numeric(task$weights$weight)
      }

      pen_optim = FALSE
      if (!is.null(opt_pars$penalty)) {
        if (opt_pars$penalty == "optimCoxBoostPenalty") {
          pen_optim = TRUE
          opt_pars$penalty = NULL
        }
      }

      with_package("CoxBoost", {
        if (pen_optim) {
          optim = mlr3misc::invoke(
            CoxBoost::optimCoxBoostPenalty,
            time = task$truth()[, 1],
            status = task$truth()[, 2],
            x = model.matrix(
              ~.,
              as.data.frame(task$data(cols = task$feature_names)))[, -1,
              drop = FALSE],
            .args = c(opt_pars, cv_pars)
          )

          return(mlr3misc::invoke(
            CoxBoost::CoxBoost,
            time = task$truth()[, 1],
            status = task$truth()[, 2],
            x = model.matrix(
              ~.,
              as.data.frame(task$data(cols = task$feature_names)))[, -1,
              drop = FALSE],
            stepno = optim$cv.res$optimal.step,
            penalty = optim$penalty,
            .args = cox_pars
          ))
        } else {
          optimal_step = mlr3misc::invoke(
            CoxBoost::cv.CoxBoost,
            time = task$truth()[, 1],
            status = task$truth()[, 2],
            x = model.matrix(
              ~.,
              as.data.frame(task$data(cols = task$feature_names)))[, -1,
              drop = FALSE],
            .args = c(cv_pars, cox_pars)
          )$optimal.step

          return(mlr3misc::invoke(
            CoxBoost::CoxBoost,
            time = task$truth()[, 1],
            status = task$truth()[, 2],
            x = model.matrix(
              ~.,
              as.data.frame(task$data(cols = task$feature_names)))[, -1,
              drop = FALSE],
            stepno = optimal_step,
            .args = cox_pars
          ))
        }
      })
    },

    .predict = function(task) {

      pars = self$param_set$get_values(tags = "predict")

      lp = as.numeric(mlr3misc::invoke(predict,
        self$model,
        newdata = model.matrix(
          ~.,
          as.data.frame(task$data(cols = task$feature_names)))[, -1,
          drop = FALSE],
        .args = pars,
        type = "lp"))

      cdf = mlr3misc::invoke(predict,
        self$model,
        newdata = model.matrix(
          ~.,
          as.data.frame(task$data(cols = task$feature_names)))[, -1,
          drop = FALSE],
        .args = pars,
        type = "CIF",
        times = sort(unique(self$model$time)))

      # define WeightedDiscrete distr6 object from predicted survival function
      x = rep(list(list(x = sort(unique(self$model$time)), cdf = 0)), task$nrow)
      for (i in 1:task$nrow) {
        x[[i]]$cdf = cdf[i, ]
      }

      distr = distr6::VectorDistribution$new(
        distribution = "WeightedDiscrete", params = x,
        decorators = c("CoreStatistics", "ExoticStatistics"))

      mlr3proba::PredictionSurv$new(task = task, crank = lp, distr = distr, lp = lp)
    }
  )
)
